From: Florian Weimer <fweimer@redhat.com>
Date: Wed, 4 Mar 2026 17:32:36 +0100
Subject: [PATCH] Linux: Only define OPEN_TREE_* macros in <sys/mount.h> if undefined

commit d12b017cddfeb9fe9920ba054ae3dfcb8e9238b8 upstream.

There is a conditional inclusion of <linux/mount.h> earlier in the file.
If that defines the macros, do not redefine them.  This addresses build
problems as the token sequence used by the UAPI macro definitions
changes between Linux versions.

For instance, the kernel headers used for the x86_64 build define
OPEN_TREE_CLONE as (1 << 0), which conflicts with the unconditional glibc
definition as 1 and aborts the build with -Werror while compiling
sysdeps/unix/sysv/linux/umount2.c:

  ../sysdeps/unix/sysv/linux/sys/mount.h:265: error: "OPEN_TREE_CLONE" redefined [-Werror]

Bug: https://sourceware.org/bugzilla/show_bug.cgi?id=33921
Reviewed-by: Adhemerval Zanella  <adhemerval.zanella@linaro.org>
Signed-off-by: Florian Weimer <fweimer@redhat.com>
[ deepin: backport to glibc 2.38 ]
Signed-off-by: hudeng <hudeng@deepin.org>
---
 sysdeps/unix/sysv/linux/sys/mount.h | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

--- a/sysdeps/unix/sysv/linux/sys/mount.h
+++ b/sysdeps/unix/sysv/linux/sys/mount.h
@@ -262,8 +262,12 @@
 #define FSOPEN_CLOEXEC          0x00000001
 
 /* open_tree flags.  */
-#define OPEN_TREE_CLONE    1         /* Clone the target tree and attach the clone */
-#define OPEN_TREE_CLOEXEC  O_CLOEXEC /* Close the file on execve() */
+#ifndef OPEN_TREE_CLONE
+# define OPEN_TREE_CLONE    1         /* Clone the target tree and attach the clone */
+#endif
+#ifndef OPEN_TREE_CLOEXEC
+# define OPEN_TREE_CLOEXEC  O_CLOEXEC /* Close the file on execve() */
+#endif
 
 
 __BEGIN_DECLS
